crates/miden-tx/src/executor/exec_host.rs

```diff
@@ -361,19 +360,27 @@ where
     /// the merkle store. Note that B' is in the store because the transaction inserted it into the
     /// merkle store as part of updating E, not because we inserted it. B is present in the store,
     /// but is simply ignored for the purpose of verifying G's inclusion.
+    ///
+    /// ## Foreign Accounts
+    ///
+    /// Foreign accounts are read-only and so they cannot change throughout transaction execution.
+    /// This means their _current_ vault root is always equivalent to their _initial_ vault root.
+    /// So, for foreign accounts, just like for the native account, we also always request
+    /// witnesses for the initial vault root.
     async fn on_account_vault_asset_witness_requested(
         &self,
         current_account_id: AccountId,
-        _vault_root: Word,
+        current_vault_root: Word,
         asset: Asset,
     ) -> Result<Vec<AdviceMutation>, TransactionKernelError> {
-        // For now, we only support getting witnesses for the native account, so return early if the
-        // requested account is not the native one.
-        if current_account_id != self.base_host.initial_account_header().id() {
-            return Ok(Vec::new());
-        }
+        // For the native account we need to explicitly request the initial vault root, while for
+        // foreign accounts the current vault root is always the initial one.
+        let vault_root = if current_account_id == self.base_host.initial_account_header().id() {
+            self.base_host.initial_account_header().vault_root()
+        } else {
+            current_vault_root
+        };
```

Comment:
Somewhat related to the above comment: could we always pass the correct root into this function? Basically, we figure out if we are dealing with the native account or not a bit more upstream, look up the correct root there, and then here, we don't need to have this `if` statement.